src/Akka.Cluster.Discovery.KubernetesApi/KubernetesDiscoveryService.cs

```diff
@@ -0,0 +1,122 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Akka.Actor;
+using Akka.Configuration;
+using Akka.Event;
+using k8s;
+using k8s.Models;
+
+namespace Akka.Cluster.Discovery.KubernetesApi
+{
+    public class KubernetesDiscoveryService : LocklessDiscoveryService
+    {
+        #region Internal classes
+
+        /// <summary>
+        /// Message scheduled by <see cref="KubernetesDiscoveryService"/> for itself. 
+        /// Used to trigger periodic restart of k8s client.
+        /// </summary>
+        public sealed class RestartClient
+        {
+            public static RestartClient Instance { get; } = new RestartClient();
+
+            private RestartClient()
+            {
+            }
+        }
+
+        #endregion
+
+        private readonly KubernetesSettings settings;
+        private IKubernetes k8s;
+        private readonly ILoggingAdapter logger = Context.GetLogger();
+        
+        private readonly string actorSystemName;
+        private readonly string labelSelector;
+
+        public KubernetesDiscoveryService(Config config) 
+            : this(new KubernetesSettings(config))
+        {            
+        }
+
+        public KubernetesDiscoveryService(KubernetesSettings settings)
+            : this(CreateKubernetesClient(settings), settings)
+        {
+        }
+
+        public KubernetesDiscoveryService(IKubernetes kubernetesClient, KubernetesSettings settings) : base(settings)
+        {
+            this.actorSystemName = ((ExtendedActorSystem)Context.System).Provider.DefaultAddress.System;
+            this.k8s = kubernetesClient;
+            this.settings = settings;
+
+            labelSelector = $"akka-cluster={actorSystemName}";
+            if (!string.IsNullOrWhiteSpace(settings.LabelSelector))
+                labelSelector = settings.LabelSelector;
+        }
+
+        private static IKubernetes CreateKubernetesClient(KubernetesSettings settings)
+        {
+            var k8sClientConfig = KubernetesClientConfiguration.InClusterConfig();
+            k8sClientConfig.Namespace = settings.KubernetesNamespace;
+            return new Kubernetes(k8sClientConfig);
+        }
+
+        private Address DeterminePodAddress(V1Pod pod)
+        {
+            if (!int.TryParse(pod.GetAnnotation("akka.remote.dot-netty.tcp.port"), out var port))
+                port = Cluster.SelfAddress.Port ?? 0;
```

Comment:
I don't think, this is a good approach. This method is used for setup the initial list of nodes to contact with. The problem here is that port `0` is a valid port only for local node, but not for any remote connection. Depending on the case this should either result in a:

1. Info saying that pod X was not annotated, so we assume current node port. In case when port is not specified throw an exception. This could however end up in weird behavior when current node was configured to use dynamic port (0), as it will be different for every OS process.
2. Warning (saying that pod X was not properly annotated with `akka.remote.dot-netty.tcp.port`) and filtering out that pod's address. I think this sounds as the best option.
2. Exception and shutdown the entire actor system.

Another issue is annotation itself: it will work only with `akka.remote.dot-netty.tcp.port`, which is the most common but not necessarily the only valid transport option. In general this should take into account the current configured transport - it can we taken from configuration using `akka.remote.enabled-transports` path (see: [here](https://github.com/akkadotnet/akka.net/blob/dev/src/core/Akka.Remote/Configuration/Remote.conf#L344)).